File: Cpp/Example4-7-PointInTriangle/Example4-7-PointInTriangle.cpp
// Example7PointInTriangle.cpp : 此文件包含 "main"
// 函数。程序执行将在此处开始并结束。
//

#include <Eigen/Eigen>
#include <iostream>

using namespace Eigen;
using namespace std;

using Triangle = Vector3d[3];

// 判断点P是否在空间三角形内
bool PointInTriangle3D(Triangle triangle, Vector3d& P) {
  auto v0p = P - triangle[0];
  auto v0v1 = triangle[1] - triangle[0];
  auto v0v2 = triangle[2] - triangle[0];

  double D = v0v1.x() * v0v2.y() - v0v1.y() * v0v2.x();
  if (D == 0.0) {
    return false;
  }

  double D1 = v0p.x() * v0v2.y() - v0p.y() * v0v2.x();
  double D2 = v0v1.x() * v0p.y() - v0v1.y() * v0p.x();

  double u = D1 / D;
  double v = D2 / D;

  //如果在三维空间判断，还需要判断第三个向量是否成立
  // double eps = v0v1.z() * u + v0v2.z() * v - P.z();

  if (u >= 0 && v >= 0 && u + v <= 1) {
    return true;
  }

  return false;
}

int main() {
  Triangle triangle = {Vector3d(-20, -25.6, 0), Vector3d(34, -24, 0),
                       Vector3d(30, 27, 0)};

  Vector3d A(1.2, 3.4, 0);
  cout << "点A在三角形内：" << PointInTriangle3D(triangle, A) << endl;

  Vector3d B(14.4, 8.14, 0);
  cout << "点B在三角形内：" << PointInTriangle3D(triangle, B) << endl;
}

File: Python/example-4-7-point-in-triangle.py
import numpy as np

class Triangle:
    def __init__(self, points):
        self.points = np.array(points)

def point_in_triangle_3d(triangle, P):
    v0p = P - triangle.points[0]
    v0v1 = triangle.points[1] - triangle.points[0]
    v0v2 = triangle.points[2] - triangle.points[0]

    D = v0v1[0] * v0v2[1] - v0v1[1] * v0v2[0]
    if D == 0.0:
        return False

    D1 = v0p[0] * v0v2[1] - v0p[1] * v0v2[0]
    D2 = v0v1[0] * v0p[1] - v0v1[1] * v0p[0]

    u = D1 / D
    v = D2 / D

    # 如果在三维空间判断，还需要判断第三个向量是否成立
    # eps = v0v1[2] * u + v0v2[2] * v - P[2]

    if u >= 0 and v >= 0 and (u + v) <= 1:
        return True

    return False

if __name__ == "__main__":
    triangle = Triangle([[-20, -25.6, 0], [34, -24, 0], [30, 27, 0]])

    A = np.array([1.2, 3.4, 0])
    print(f"点A在三角形内：{point_in_triangle_3d(triangle, A)}")

    B = np.array([14.4, 8.14, 0])
    print(f"点B在三角形内：{point_in_triangle_3d(triangle, B)}")